### ISSUE TITLE
Hanging transfer

### ISSUE THREAD

**Issue by kelsos** [raiden-network/raiden#3146]
## Problem Definition

I don't have complete information about the network topology but I would guess it doesn't really matter.

There are two important nodes in this case:

Node A: `0x0E809A051034723beE67871a5A4968aE22d36C5A` (Initiator)
Node B: `0x53A9462Be18D8f74C1065Be65A58D5A41347e0A6` (Mediator)
...
Node X: `0x8f476EaE54684ABDc6f2206ef903064956fAAbb0` (Receiver)

`A` Was sending sending 200 transfers, of 1 wei each `X` through `B`. Somewhere mid transfer `B` was restarted in order to check if the payment would be routed through another route in the network.

The transfers where indeed routed through an other node, but during the test after restarting the node @agatsoh reported that he was unable to send transfers to another node, that was connected to his node through the `B`->`A` path.

After some investigation it seems that after the restart `B` could not send anything back to `A` and a had `Nonce did not change sequentially` errors in the logs.

You can find the complete [logs & databases](https://drive.google.com/open?id=1VfC29Mv7gXku69d1eZ9b0dRcDMeCcJ6K)

### System Description

Here add a detailed description of your system, e.g. output of the following script:

```
{
  "raiden": "0.18.1",
  "python_implementation": "CPython",
  "python_version": "3.6.7",
  "system": "Linux 64bit_ELF 4.19.4-arch1-1-ARCH x86_64",
  "distribution": "bundled"
}

Linux forseti 4.19.4-arch1-1-ARCH #1 SMP PREEMPT Fri Nov 23 09:06:58 UTC 2018 x86_64 GNU/Linux
/usr/bin/solc
solc, the solidity compiler commandline interface
Version: 0.5.0+commit.1d4f565a.Linux.g++
/usr/bin/geth
Geth
Version: 1.8.19-stable
Git Commit: dae82f098570e15d44584f0d7f350713f4774727
Architecture: amd64
Protocol Versions: [63 62]
Network Id: 1
Go Version: go1.11.2
Operating System: linux
GOPATH=
GOROOT=/usr/lib/go
/usr/bin/parity
Parity Ethereum
  version Parity-Ethereum/v2.1.8-stable-3eae1d3640-20181130/x86_64-linux-gnu/rustc1.30.1
Copyright 2015-2018 Parity Technologies (UK) Ltd.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

By Wood/Paronyan/Kotewicz/Drwięga/Volf
   Habermeier/Czaban/Greeff/Gotchac/Redmann

```

**Comment by LefterisJP** [comment]
I am looking at the logs now a bit to get a bit more general idea of what's happening. 

Sidenote: One request would be, if possible, for people who are testing to send us an as small as possible reproducible test case with small logs. Looking through all the logs in this issue is really hard and even makes my firefox hang.

**Comment by LefterisJP** [comment]
So some first results of investigating. I started by looking at the logs of `0x0E809A051034723beE67871a5A4968aE22d36C5A` who experienced the "Nonce Mismatch" problem.

Specifically I used this log and command to generate the html output:

`python tools/debugging/json-log-to-html.py -o ~/3146-0x0E8-11-12-09-12.html ~/3146_logs/0x0E809A051034723beE67871a5A4968aE22d36C5A/raiden-debug_2018-12-11T09:12:36.324165.log -l filelock -l raiden.api.rest.pywsgi`

First bad message is a locked transfer with nonce `199` while we expect `198` at line 25870. 
- initiator: 0x53A9462Be18D8f74C1065Be65A58D5A41347e0A6
- secrethash: 0x33f25f0c3a6a9234ba158b842f144b6d239c02d2e3ff72a5c545f41d3b998486
- channel identifier: 137
- nonce:199

```
node=0e809a05 greenlet_name=sync_handle_response-s1032883_3498787_0_1_1054_1_1_6847_1 state_change={"route": {"node_address": "0x53A9462Be18D8f74C1065Be65A58D5A41347e0A6", "channel_identifier": "137", "_type": "raiden.transfer.state.RouteState", "_version": 0}, "transfer": {"message_identifier": "6517512766869490903", "payment_identifier": "1544526490308", "token": "0xd0A1E359811322d97991E03f863a0C30C2cF029C", "balance_proof": {"nonce": 199, "transferred_amount": "271", "locked_amount": "1", "locksroot": "0xc6bec5e45486c21090d9417a7c44f4c9838ec97be4ee7b17e7058f8b4a110ec4", "token_network_identifier": "0x111157460c0F41EfD9107239B7864c062aA8B978", "channel_identifier": "137", "message_hash": "0x01c8789c0ad1214f641e02b7a24ce61447416ed8c68197878aa18354a0529dfe", "signature": "0x217da8b21c3a1e866f0c0bc6bf01b43ada16ae161c61b7b01256906b483ec5663fc2af22774f63c17cc84d3e3258181a7eb80ba52d2d5c4de32e1636442a2efc1b", "sender": "0x53A9462Be18D8f74C1065Be65A58D5A41347e0A6", "chain_id": 42, "balance_hash": "0x56bd6c8185319f8c729ed1f3c7eaeaacd41854d65d4ac60540d7cb076cf97e00", "_type": "raiden.transfer.state.BalanceProofSignedState", "_version": 0}, "lock": {"amount": 1, "expiration": 9662495, "secrethash": "0x33f25f0c3a6a9234ba158b842f144b6d239c02d2e3ff72a5c545f41d3b998486", "_type": "raiden.transfer.state.HashTimeLockState", "_version": 0}, "initiator": "0x53A9462Be18D8f74C1065Be65A58D5A41347e0A6", "target": "0x0E809A051034723beE67871a5A4968aE22d36C5A", "_type": "raiden.transfer.mediated_transfer.state.LockedTransferSignedState", "_version": 0}, "balance_proof": {"nonce": 199, "transferred_amount": "271", "locked_amount": "1", "locksroot": "0xc6bec5e45486c21090d9417a7c44f4c9838ec97be4ee7b17e7058f8b4a110ec4", "token_network_identifier": "0x111157460c0F41EfD9107239B7864c062aA8B978", "channel_identifier": "137", "message_hash": "0x01c8789c0ad1214f641e02b7a24ce61447416ed8c68197878aa18354a0529dfe", "signature": "0x217da8b21c3a1e866f0c0bc6bf01b43ada16ae161c61b7b01256906b483ec5663fc2af22774f63c17cc84d3e3258181a7eb80ba52d2d5c4de32e1636442a2efc1b", "sender": "0x53A9462Be18D8f74C1065Be65A58D5A41347e0A6", "chain_id": 42, "balance_hash": "0x56bd6c8185319f8c729ed1f3c7eaeaacd41854d65d4ac60540d7cb076cf97e00", "_type": "raiden.transfer.state.BalanceProofSignedState", "_version": 0}, "_type": "raiden.transfer.mediated_transfer.state_change.ActionInitTarget", "_version": 0}
```

Then looking back trying to see if we ever get `198` we can see that in log line 103 we get the nonce `198` for that channel for the first time. And it's a `ReceiveLockExpired` message.

```
node=0e809a05 greenlet_name=sync_handle_response-s995981_3491877_0_1_1034_1_1_6808_1 state_change={"balance_proof": {"nonce": 198, "transferred_amount": "271", "locked_amount": "0", "locksroot": "0x0000000000000000000000000000000000000000000000000000000000000000", "token_network_identifier": "0x111157460c0F41EfD9107239B7864c062aA8B978", "channel_identifier": "137", "message_hash": "0xf3c4f6016cd30fbe2e74b40658179416199bd874063bfe14017b066bd06ef422", "signature": "0xb69d77150011bdd3e9655347dfaf61c782241f6cf329a2a6e22ca0ca5934b1851a25da06effc6b7f805dcba8778a64c397d1c9af913fb3be8a2ad3274617d6041b", "sender": "0x53A9462Be18D8f74C1065Be65A58D5A41347e0A6", "chain_id": 42, "balance_hash": "0x598f39c712639e9fd2e2c25edf630054e9f3665bc930c3579fdf5e26128846ec", "_type": "raiden.transfer.state.BalanceProofSignedState", "_version": 0}, "secrethash": "0xe17f5b8438379c15e6b684fd5f03c4ceedd1b8981966fc742df543e46feae233", "message_identifier": "3404614811605573059", "_type": "raiden.transfer.mediated_transfer.state_change.ReceiveLockExpired", "_version": 0}
```

So we actually do get the message with nonce `198`. In the next line we send a `Delivered` for it to the sender. But then in the following lines no Raiden events are generated. Looking at [target](https://github.com/raiden-network/raiden/blob/8bb02c11dd863930a1f5dfbed0e417eec784d925/raiden/transfer/mediated_transfer/target.py#L286) processing of that message and subsequently at the [channel](https://github.com/raiden-network/raiden/blob/8bb02c11dd863930a1f5dfbed0e417eec784d925/raiden/transfer/channel.py#L1508) processing we can see that we always generate an event. We either [send a processed](https://github.com/raiden-network/raiden/blob/8bb02c11dd863930a1f5dfbed0e417eec784d925/raiden/transfer/channel.py#L1534) if all is fine or emit an [invalid lock expired](https://github.com/raiden-network/raiden/blob/8bb02c11dd863930a1f5dfbed0e417eec784d925/raiden/transfer/channel.py#L1540) message if the `LockExpired` message is not valid.

Neither of the two seems to be happening. Perhaps figuring out why will help towards solving this issue.

**Comment by LefterisJP** [comment]
Sidetracked a bit due to this:

> For that LockExpired (and perhaps for all of them) we don't get any events as described above. But we do get the EventUnlockClaimFailed event as above. And what's interesting is that it's the same exact event.
node=0e809a05 greenlet_name=AlarmTask\|Greenlet-2 raiden_event={"identifier":  "1544452131", "secrethash":  "0x0852485649626d28c1e602c2e0787d108ad49321c35438cf4ae1f83140badb3e",  "reason": "lock expired", "_type":  "raiden.transfer.mediated_transfer.events.EventUnlockClaimFailed",  "_version": 0}

It seems that in the logs of `0x0E80..` it's full of `EventUnlockClaimFailed` for the same transfer as can be seen above. Turns out it was the target of this transfer and as such we still had https://github.com/raiden-network/raiden/issues/3086 but for the target's side. I addressed it just now in a PR: https://github.com/raiden-network/raiden/pull/3150

Note though that this has nothing to do with the original problem. Just realized another problem and fixed it after looking through the logs.

**Comment by kelsos** [comment]
So I manage to get the following reproducible scenario. At least I manage to get to the exact same situation with the exact same steps 3 times in a row. 

You can see the diagram below:

![nonce_scenario](https://user-images.githubusercontent.com/2269732/49935636-d9c53480-fed1-11e8-8d88-9d068303dc95.png)

`0x82641569b2062B545431cF6D7F0A418582865ba7` and `0x60FFe4daDE358246C49C4A16bB5b430d333B5Ce9` are sending 200 transfers of 1 to each other.

At some point `0xc52952Ebad56f2c5E5b42bb881481Ae27D036475` that mediated the transfers goes offline and stays offline until the Locks on the transfers on both directions expire (and the REST API returns 409 on both the senders). Then `0xc52952E` gets restarted and  `0x8264156` gives `Nonce did not change sequentially` errors when receiving anything from `0x504300C`. 

At this point any transfers from `0x504300C` to `0x8264156` will hang.

While I can't be 100% sure that this is the exact same situation we faced testing, it is rather similar to the actual test network since Cosmin had connections to all of my channels and we restarted his nodes to check if the transfers would get properly rerouted through other nodes.

**Comment by LefterisJP** [comment]
> So we actually do get the message with nonce 198. In the next line we send a Delivered for it to the sender. But then in the following lines no Raiden events are generated. Looking at target processing of that message and subsequently at the channel processing we can see that we always generate an event. We either send a processed if all is fine or emit an invalid lock expired message if the LockExpired message is not valid.
>
>Neither of the two seems to be happening. We only get an EventUnlockClaimFailed for a different secrethash.

@rakanalh and myself spent some time debugging this by replaying the DB in the Berlin office and it seems that the reason no events are dispatched is because there is no longer a [payment subtask](https://github.com/raiden-network/raiden/blob/76b8456dd564bf0b374ddb4fc6539dbde1bb664b/raiden/transfer/node.py#L145) to dispatch to.

Looking at mediatior's `clear_if_finalized()`

https://github.com/raiden-network/raiden/blob/76b8456dd564bf0b374ddb4fc6539dbde1bb664b/raiden/transfer/mediated_transfer/mediator.py#L276-L293

that can only happen if either both payer and payee channels are gone or if there is no pending locks for payer channel's partner state and payee channel's our state.

So now we are gonna look into why the task was removed and/or why was the lock expired sent after the task was removed. 

An interesting thing to note is that before all these things happened there was also a refund transfer sent from the node.

**Comment by LefterisJP** [comment]
> These are the logs from one of the reproduced cases.

So @rakanalh and myself have been looking at the logs and DBs for 2 days and have found at least 2 different problems. Up to the sideffect of the point described in my [comment](https://github.com/raiden-network/raiden/issues/3146#issuecomment-447027876) above both problems appear the same.

 I will describe here the problem is see with [these](https://drive.google.com/open?id=1pn1HSRZiV9ITMi9scpPGiG_KeSq1QECj) logs I quoted from above. @rakanalh can perhaps describe the other problem in another issue related to this so that we have 2 different changelog entries.

---

## Logs of 82641569

First Nonce did not change sequentially message is "expected is 7 got 9"
The balance proof there contains the following identifying attributes:
`channel identifier`: `3`
`sender`: `0x504300C525CbE91Adb3FE0944Fe1f56f5162C75C`
`target`: `0x60FFe4daDE358246C49C4A16bB5b430d333B5Ce9`

Searching the DB for the nonce `7` message and we see it's a `LockExpired` and no events are generated. Its secrethash is `b'\xd2\xea\xb0\xef\x00\x97\xf8\xab\x91\x02\xbauv\xd2\xd0\xe8\xf1b\xf2\xb8\x064\x16>\x8e\xf4\xe1\x14\xe8N\x0e\xa1'`

The first time this `secrethash` is seen in the logs is:
`<ActionInitInitiator transfer:<TransferDescriptionWithSecretState token_network:15008736 amount:1 target:60ffe4da secrethash:d2eab0ef>>`

Then the following happen:

```
<ReceiveTransferRefundCancelRoute sender:504300c5 transfer:<LockedTransferSignedState msgid:1097169334591442817 id:1544785585 token:0x28104ee15e1c70c421150865c3fb731c426929e6 lock:<HashTimeLockState amount:1 expiration:715 secrethash:d2eab0ef> target:0x60ffe4dade358246c49c4a16bb5b430d333b5ce9>>
----  <SendProcessed confirmed_msgid:1097169334591442817 recipient:504300c5>
----  <EventUnlockFailed id:1544785585 secrethash:5374ee94 reason:route was canceled>
----  <EventPaymentSentFailed payment_network_identifier:a4e13d32 token_network_identifier:15008736 id:1544785585 target:b'`\xff\xe4\xda\xde5\x82F\xc4\x9cJ\x16\xbb[C\r3;\\\xe9' reason:there is no route available >
```

And after that we get the Lock Expired with nonce == 7

In our current codebase canceling a route without any other route existing will destroy the payment task. This is important to remember.

## Logs of 504300c5

First time the `secrethash` is seen is when it tries to send a refund transfer back to `82641569`

`<SendRefundTransfer msgid:1097169334591442817 transfer:<LockedTransferUnsignedState id:1544785585 token:0x28104ee15e1c70c421150865c3fb731c426929e6 balance_proof:<BalanceProofUnsignedState nonce:3 transferred_amount:1 locked_amount:1 locksroot:1368f4ce token_network:15008736 channel_identifier:3 chain_id:17> lock:<HashTimeLockState amount:1 expiration:715 secrethash:d2eab0ef> target:0x60ffe4dade358246c49c4a16bb5b430d333b5ce9> recipient:82641569 >`

Once this is done a lot of things happen.

```
<ActionInitMediator from_route:<RouteState hop:82641569 channel_identifier:3> from_transfer:<LockedTransferSignedState msgid:6403070057224337511 id:1544785616 token:0x28104ee15e1c70c421150865c3fb731c426929e6 lock:<HashTimeLockState amount:1 expiration:744 secrethash:2051de9c> target:0x60ffe4dade358246c49c4a16bb5b430d333b5ce9>>
----  <SendProcessed confirmed_msgid:6403070057224337511 recipient:82641569>
----  <SendRefundTransfer msgid:8431740765894415002 transfer:<LockedTransferUnsignedState id:1544785616 token:0x28104ee15e1c70c421150865c3fb731c426929e6 balance_proof:<BalanceProofUnsignedState nonce:4 transferred_amount:1 locked_amount:2 locksroot:4ef4594f token_network:15008736 channel_identifier:3 chain_id:17> lock:<HashTimeLockState amount:1 expiration:744 secrethash:2051de9c> target:0x60ffe4dade358246c49c4a16bb5b430d333b5ce9> recipient:82641569 >
<ActionInitTarget route:<RouteState hop:82641569 channel_identifier:3> transfer:<LockedTransferSignedState msgid:2369951105086519663 id:1544785658 token:0x28104ee15e1c70c421150865c3fb731c426929e6 lock:<HashTimeLockState amount:1 expiration:784 secrethash:9f99e9ca> target:0x504300c525cbe91adb3fe0944fe1f56f5162c75c>>
----  <SendProcessed confirmed_msgid:2369951105086519663 recipient:82641569>
----  <SendSecretRequest msgid:12465429875095018829 paymentid:1544785658 amount:1 expiration:784 secrethash:9f99e9ca recipient:82641569>
<ActionInitInitiator transfer:<TransferDescriptionWithSecretState token_network:15008736 amount:1 target:82641569 secrethash:f1c537ae>>
----  <SendLockedTransfer msgid:5360692394182293743 transfer:<LockedTransferUnsignedState id:1544785659 token:0x28104ee15e1c70c421150865c3fb731c426929e6 balance_proof:<BalanceProofUnsignedState nonce:5 transferred_amount:1 locked_amount:3 locksroot:3774dc41 token_network:15008736 channel_identifier:3 chain_id:17> lock:<HashTimeLockState amount:1 expiration:785 secrethash:f1c537ae> target:0x82641569b2062b545431cf6d7f0a418582865ba7> recipient:82641569>
<ReceiveUnlock msgid:10439263912960393569 secrethash:9f99e9ca balance_proof:<BalanceProofSignedState nonce:6 transferred_amount:2 locked_amount:2 locksroot:4ef4594f token_network:15008736 channel_identifier:3 message_hash:2ad6f6e3 signature:358e35b8 sender:82641569 chain_id:17>>
----  <SendProcessed confirmed_msgid:10439263912960393569 recipient:82641569>
----  <EventPaymentReceivedSuccess payment_network_identifier:a4e13d32 token_network_identifier:15008736 identifier:1544785658 amount:1 initiator:82641569 >
----  <EventUnlockClaimSuccess id:1544785658 secrethash:9f99e9ca>
----  <SendProcessed confirmed_msgid:10439263912960393569 recipient:82641569>
<ActionInitInitiator transfer:<TransferDescriptionWithSecretState token_network:15008736 amount:1 target:c52952eb secrethash:c271256b>>
----  <SendLockedTransfer msgid:7765947373046617195 transfer:<LockedTransferUnsignedState id:1544785660 token:0x28104ee15e1c70c421150865c3fb731c426929e6 balance_proof:<BalanceProofUnsignedState nonce:3 transferred_amount:1 locked_amount:1 locksroot:69d69b8f token_network:15008736 channel_identifier:4 chain_id:17> lock:<HashTimeLockState amount:1 expiration:786 secrethash:c271256b> target:0xc52952ebad56f2c5e5b42bb881481ae27d036475> recipient:c52952eb>
<ActionInitTarget route:<RouteState hop:c52952eb channel_identifier:4> transfer:<LockedTransferSignedState msgid:7812258135248903461 id:1544785664 token:0x28104ee15e1c70c421150865c3fb731c426929e6 lock:<HashTimeLockState amount:1 expiration:790 secrethash:c3a0e76e> target:0x504300c525cbe91adb3fe0944fe1f56f5162c75c>>
----  <SendProcessed confirmed_msgid:7812258135248903461 recipient:c52952eb>
----  <SendSecretRequest msgid:17298645885062014750 paymentid:1544785664 amount:1 expiration:790 secrethash:c3a0e76e recipient:c52952eb>
```

The important thing from the above is that `504300c5` sends a RefundTransfer to `82641569`

After all the above some time passes and the lock expiration block of that secrethash comes and the lock expires:

```
(Pdb++) p state_change
<Block number=725 gas_limit=8000000 block_hash=4046b326>
(Pdb++) p event
<EventUnlockClaimFailed id:1544785585 secrethash:d2eab0ef reason:lock expired>
(Pdb++)p event 
<SendLockExpired msgid:8608338305876032644 balance_proof:<BalanceProofUnsignedState nonce:7 transferred_amount:2 locked_amount:1 locksroot:4b0dc3e6 token_network:15008736 channel_identifier:3 chain_id:17> secrethash:b'\xd2\xea\xb0\xef\x00\x97\xf8\xab\x91\x02\xbauv\xd2\xd0\xe8\xf1b\xf2\xb8\x064\x16>\x8e\xf4\xe1\x14\xe8N\x0e\xa1' recipient:82641569>
(Pdb++)p event 
<EventUnlockFailed id:1544785585 secrethash:d2eab0ef reason:lock expired>
```

When that happens `504300c5`, as we see above sends a `SendLockExpired` message to  `82641569`. And that's what causes the problem of the nonce mismatch because as mentioned in the comment [above](https://github.com/raiden-network/raiden/issues/3146#issuecomment-447027876) the initiator `82641569` no longer has a payment task for that secrethash and generates no events, so essentially does not process the lock expired message.

---

Solution is coming in a PR.
